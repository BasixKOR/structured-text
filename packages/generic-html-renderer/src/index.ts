import {
  Adapter,
  isBlockquote,
  isCode,
  isHeading,
  isLink,
  isList,
  isListItem,
  isParagraph,
  isRoot,
  isSpan,
  Mark,
  Record,
  render as genericRender,
  RenderResult,
  RenderRule,
  renderRule,
  StructuredText,
  TrasformFn,
  RenderError,
} from 'datocms-structured-text-utils';

export { renderRule, RenderError };

export function markToTagName(mark: Mark): string {
  switch (mark) {
    case 'emphasis':
      return 'em';
    case 'underline':
      return 'u';
    case 'strikethrough':
      return 'del';
    case 'highlight':
      return 'mark';
    default:
      return mark;
  }
}

export function render<
  R extends Record,
  H extends TrasformFn,
  T extends TrasformFn,
  M extends TrasformFn,
  F extends TrasformFn
>(
  adapter: Adapter<H, T, M, F>,
  structuredText: StructuredText<R>,
  customRules: RenderRule<H, T, M, F>[],
): RenderResult<H, T, M, F> {
  return genericRender(adapter, structuredText, [
    ...customRules,
    renderRule(isRoot, ({ adapter: { renderFragment }, key, children }) => {
      return renderFragment(children, key);
    }),
    renderRule(isParagraph, ({ adapter: { renderNode }, key, children }) => {
      return renderNode('p', { key }, children);
    }),
    renderRule(isList, ({ adapter: { renderNode }, node, key, children }) => {
      return renderNode(
        node.style === 'bulleted' ? 'ul' : 'ol',
        { key },
        children,
      );
    }),
    renderRule(isListItem, ({ adapter: { renderNode }, key, children }) => {
      return renderNode('li', { key }, children);
    }),
    renderRule(isBlockquote, ({ adapter: { renderNode }, key, children }) => {
      return renderNode('blockquote', { key }, children);
    }),
    renderRule(isCode, ({ adapter: { renderNode, renderText }, key, node }) => {
      return renderNode(
        'pre',
        { key, 'data-language': node.language },
        renderNode('code', null, renderText(node.code)),
      );
    }),
    renderRule(isLink, ({ adapter: { renderNode }, key, children, node }) => {
      return renderNode('a', { key, href: node.url }, children);
    }),
    renderRule(
      isHeading,
      ({ node, adapter: { renderNode }, children, key }) => {
        return renderNode(`h${node.level}`, { key }, children);
      },
    ),
    renderRule(isSpan, ({ adapter: { renderNode, renderText }, key, node }) => {
      const marks = node.marks || [];

      const lines = node.value.split(/\n/);

      const textWithNewlinesConvertedToBr =
        lines.length > 0
          ? lines.slice(1).reduce(
              (acc, line, index) => {
                return acc.concat([
                  renderNode('br', { key: `br-${index}` }),
                  renderText(line, `line-${index}`),
                ]);
              },
              [renderText(lines[0], 'line-first')],
            )
          : renderText(node.value, key);

      return marks.reduce<RenderResult<H, T, M, F>>((children, mark) => {
        return renderNode(markToTagName(mark), { key }, children);
      }, textWithNewlinesConvertedToBr);
    }),
  ]);
}
