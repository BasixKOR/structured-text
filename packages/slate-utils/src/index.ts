export * from './slateToDast';
export * from './types';
export * from './guards';
