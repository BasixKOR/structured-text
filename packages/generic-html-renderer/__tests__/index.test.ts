import {
  isBlock,
  StructuredText,
  isInlineItem,
} from 'datocms-structured-text-utils';
import { isItemLink } from '../../utils/src';
import { render, renderNodeRule, renderMarkRule } from '../src';

type ImageRecord = {
  id: string;
  __typename: 'ImageRecord';
  image: {
    url: string;
  };
};

type BlogPostRecord = {
  id: string;
  __typename: 'BlogPostRecord';
  title: string;
  author: {
    name: string;
  };
};

type Tag = {
  tagName: string;
  attrs: Record<string, string>;
  children: Array<
    undefined | Text | Fragment | Tag | Array<Text | Fragment | Tag>
  >;
};

type Text = {
  text: string;
  key: string;
};

type Fragment = {
  children: Array<Tag | Text | Fragment> | undefined;
  key: string;
};

const dummyAdapter = {
  renderNode: (
    tagName: string,
    attrs: Record<string, string>,
    ...children: Array<
      undefined | Text | Fragment | Tag | Array<Text | Fragment | Tag>
    >
  ): Tag => ({
    tagName,
    attrs,
    children,
  }),
  renderFragment: (
    children: (Tag | Text | Fragment)[],
    key: string,
  ): Fragment => ({
    children,
    key,
  }),
  renderText: (text: string, key: string): Text => ({ text, key }),
};

const renderRulesForValue = (
  value: StructuredText<ImageRecord | BlogPostRecord>,
) => [
  renderNodeRule(
    isInlineItem,
    ({ node, adapter: { renderNode, renderText }, key }) => {
      const record = value.links?.find((record) => record.id === node.item);

      if (!record) {
        return null;
      }

      switch (record.__typename) {
        case 'BlogPostRecord':
          return renderNode('a', { key, href: `/blog/${record.id}` }, [
            renderText(record.title, 'p-0'),
          ]);
        default:
          return null;
      }
    },
  ),
  renderNodeRule(
    isItemLink,
    ({ node, adapter: { renderNode }, key, children }) => {
      const record = value.links?.find((record) => record.id === node.item);

      if (!record) {
        return null;
      }

      switch (record.__typename) {
        case 'BlogPostRecord':
          return renderNode('a', { key, href: `/blog/${record.id}` }, children);
        default:
          return null;
      }
    },
  ),
  renderNodeRule(isBlock, ({ node, adapter: { renderNode }, key }) => {
    const record = value.blocks?.find((record) => record.id === node.item);

    if (!record) {
      return null;
    }

    switch (record.__typename) {
      case 'ImageRecord':
        return renderNode('img', { key, src: record.image.url });
      default:
        return null;
    }
  }),
];

describe('render', () => {
  it('transforms a structured text GraphQL response into HTML using an adapter', () => {
    const value: StructuredText<ImageRecord | BlogPostRecord> = {
      value: {
        schema: 'dast',
        document: {
          type: 'root',
          children: [
            {
              type: 'heading',
              level: 1,
              children: [
                {
                  type: 'span',
                  value: 'This is a ',
                },
                {
                  type: 'span',
                  marks: ['highlight'],
                  value: 'title',
                },
                {
                  type: 'span',
                  value: '.',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value: "And it's berry perfect.",
                },
              ],
            },
            {
              type: 'heading',
              level: 2,
              children: [
                {
                  type: 'span',
                  value: 'Seriously, what the grape else do you want?',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    'You probably build websites and think your banana is special.\nYou think your 13 megabyte parallax-ative home page is going to get you',
                },
                {
                  type: 'span',
                  marks: ['strong'],
                  value: ' some fruity Awwward banner y',
                },
                {
                  type: 'span',
                  value:
                    'ou can glue to the top corner of your site. You think your 40-pound jQuery file and 83 polyfills give IE7 a peach because it finally has box-shadow. Wrong, melonhead. Let me describe your perfect-pear website:',
                },
              ],
            },
            {
              type: 'list',
              style: 'bulleted',
              children: [
                {
                  type: 'listItem',
                  children: [
                    {
                      type: 'paragraph',
                      children: [
                        {
                          type: 'span',
                          value: "Fruit's lightweight and loads fast",
                        },
                      ],
                    },
                    {
                      type: 'list',
                      style: 'bulleted',
                      children: [
                        {
                          type: 'listItem',
                          children: [
                            {
                              type: 'paragraph',
                              children: [
                                {
                                  type: 'span',
                                  value:
                                    'Looks the same in all your citrusy browsers',
                                },
                              ],
                            },
                          ],
                        },
                        {
                          type: 'listItem',
                          children: [
                            {
                              type: 'paragraph',
                              children: [
                                {
                                  type: 'span',
                                  value: 'Fits on all your juicy screens',
                                },
                              ],
                            },
                            {
                              type: 'list',
                              style: 'bulleted',
                              children: [
                                {
                                  type: 'listItem',
                                  children: [
                                    {
                                      type: 'paragraph',
                                      children: [
                                        {
                                          type: 'span',
                                          value:
                                            "The kumquat's accessible to every apricot that visits your site",
                                        },
                                      ],
                                    },
                                  ],
                                },
                              ],
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
                {
                  type: 'listItem',
                  children: [
                    {
                      type: 'paragraph',
                      children: [
                        {
                          type: 'span',
                          value:
                            "Fruit's legible and gets your juicy point across (if you had one instead of just 5mb pics of hipsters drinking coffee)",
                        },
                      ],
                    },
                  ],
                },
              ],
            },
            {
              type: 'heading',
              level: 3,
              children: [
                {
                  type: 'span',
                  value: 'Well guess what, you absolute kiwi:',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    "You. Are. Over-designing. Look at this fruit. It's a banana-loving website. Why the fig do you need to animate a fruity trendy-pear banner flag when I hover over that useless piece of mango? You spent hours on it and added 80 kilobytes to your juicy site, and some papaya jabbing at it on their iPad with fat sausage fingers will never see that fruit. Not to mention blind people will never see that fruit, but they don't see any of your juicy fruit.",
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    "You never knew it, but this is your perfect website. Here's why.",
                },
              ],
            },
            {
              type: 'heading',
              level: 2,
              children: [
                {
                  type: 'span',
                  value: "It's fruity lightweight",
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    'This entire page weighs less than the gradient-meshed facebook logo on your juicy Wordpress site. Did you seriously load 100kb of jQuery UI just so you could animate the fruity background color of a div? You loaded all 7 fontfaces of a citrusy webfont just so you could say "Hi." at 100px height at the beginning of your site? ',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value: 'You piece of fruit.',
                },
              ],
            },
            {
              type: 'list',
              style: 'bulleted',
              children: [
                {
                  type: 'listItem',
                  children: [
                    {
                      type: 'paragraph',
                      children: [
                        {
                          item: '11924676',
                          type: 'inlineItem',
                        },
                        {
                          item: '11924676',
                          type: 'itemLink',
                          children: [
                            {
                              type: 'span',
                              value: 'Foobar',
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
                {
                  type: 'listItem',
                  children: [
                    {
                      type: 'paragraph',
                      children: [
                        {
                          type: 'span',
                          value: '',
                        },
                      ],
                    },
                  ],
                },
              ],
            },
            {
              item: '11922283',
              type: 'block',
            },
            {
              type: 'heading',
              level: 2,
              children: [
                {
                  type: 'span',
                  value: "It's responsive",
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    "You kumquat. You thought you needed media queries to be responsive, but no. Responsive means that it responds to whatever fruity screensize it's viewed on. This site doesn't care if you're on an iMac or a banana-loving Tamagotchi.",
                },
              ],
            },
            {
              type: 'heading',
              level: 2,
              children: [
                {
                  type: 'span',
                  value: 'It fruity works',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    "Look at this fruit. You can read it ... that is, if you can read, melonhead. It makes sense. It has banana-loving hierarchy. It's using HTML5 tags so you and your pear-headed browser know what the fig's in this juicy site. That's semantics, you absolute kiwi.",
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    'It has content on the fruity screen. Your site has three bylines and link to your dribbble account, but you spread it over 7 full screens and make me click some bobbing button to show me how cool the jQuery ScrollTo plugin is.',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    'Cross-browser compatibility? Load this kumquat in IE6. I fruity dare you.',
                },
              ],
            },
            {
              type: 'heading',
              level: 2,
              children: [
                {
                  type: 'span',
                  value: 'This is a website. ',
                },
                {
                  type: 'span',
                  marks: ['highlight'],
                  value: 'Look at it.',
                },
                {
                  type: 'span',
                  value: " You've never seen one before.",
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    "Like the man who's never grown out his beard has no idea what his true natural state is, you have no fruity idea what a website is. All you have ever seen are citrusy skeuomorphic bastardizations of what should be text communicating a juicy message. This is a real, naked website. Look at it. It's fruity beautiful.",
                },
              ],
            },
            {
              type: 'heading',
              level: 3,
              children: [
                {
                  type: 'span',
                  value: 'Yes, this is fruity satire, you fig',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value:
                    "I'm not actually saying your citrusy site should look like this. What I'm saying is that all the problems we have with websites are ones we create ourselves. Websites aren't broken by default, they are functional, high-performing, and accessible. You break them. You son-of-a-peach.",
                },
              ],
            },
            {
              type: 'blockquote',
              attribution: 'some German fruithead',
              children: [
                {
                  type: 'paragraph',
                  children: [
                    {
                      type: 'span',
                      value: '"Good design is as little design as possible." ',
                    },
                  ],
                },
              ],
            },
            {
              type: 'heading',
              level: 2,
              children: [
                {
                  type: 'span',
                  value: 'Epilogue',
                },
              ],
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'span',
                  value: 'From the philosophies expressed (poorly) above, ',
                },
                {
                  url: 'http://txti.es',
                  type: 'link',
                  meta: [
                    { id: 'rel', value: 'nofollow' },
                    { id: 'foo', value: '123' },
                    { id: 'target', value: '_blank' },
                  ],
                  children: [
                    {
                      type: 'span',
                      value: 'txti',
                    },
                  ],
                },
                {
                  type: 'span',
                  value:
                    ' was created. You should try it today to make your own banana-loving websites.',
                },
              ],
            },
          ],
        },
      },
      blocks: [
        {
          __typename: 'ImageRecord',
          id: '11922283',
          image: {
            url:
              'https://www.datocms-assets.com/205/1606492977-oberlo-cover.jpg',
          },
        },
      ],
      links: [
        {
          __typename: 'BlogPostRecord',
          id: '11924676',
          title: 'New feature: Media Area granular permissions',
          author: {
            name: 'Stefano Verna',
          },
        },
      ],
    };

    expect(
      render(value, {
        adapter: dummyAdapter,
        customNodeRules: renderRulesForValue(value),
      }),
    ).toMatchSnapshot();
  });

  it('unique keys', () => {
    const value: StructuredText<ImageRecord | BlogPostRecord> = {
      value: {
        schema: 'dast',
        document: {
          type: 'root',
          children: [
            {
              type: 'paragraph',
              children: [
                { type: 'span', value: 'text two\\n' },
                { type: 'span', value: 'text two\\n' },
              ],
            },
          ],
        },
      },
    };

    expect(
      render(value, {
        adapter: dummyAdapter,
        customNodeRules: renderRulesForValue(value),
      }),
    ).toMatchSnapshot();
  });

  it('custom span renderer', () => {
    const value: StructuredText<ImageRecord | BlogPostRecord> = {
      value: {
        schema: 'dast',
        document: {
          type: 'root',
          children: [
            {
              type: 'paragraph',
              children: [
                { type: 'span', marks: ['emphasis'], value: 'text\ntwo' },
              ],
            },
          ],
        },
      },
    };

    expect(
      render(value, {
        adapter: dummyAdapter,
        customMarkRules: [
          renderMarkRule(
            'emphasis',
            ({ adapter: { renderNode }, key, children }) =>
              renderNode('something-different', { key }, children),
          ),
        ],
      }),
    ).toMatchSnapshot();
  });
});
