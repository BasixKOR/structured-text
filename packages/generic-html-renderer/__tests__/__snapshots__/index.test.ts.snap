// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render transforms a structured text GraphQL response into HTML using an adapter 1`] = `
Object {
  "children": Array [
    Object {
      "attrs": Object {
        "key": "t-0",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "This is a ",
              },
            ],
            "key": "t-0",
          },
          Object {
            "attrs": Object {
              "key": "t-1",
            },
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "key": "t-1-line-first",
                    "text": "title",
                  },
                ],
                "key": "t-1",
              },
            ],
            "tagName": "mark",
          },
          Object {
            "children": Array [
              Object {
                "key": "t-2-line-first",
                "text": ".",
              },
            ],
            "key": "t-2",
          },
        ],
      ],
      "tagName": "h1",
    },
    Object {
      "attrs": Object {
        "key": "t-1",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "And it's fucking perfect.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-2",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Seriously, what the fuck else do you want?",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h2",
    },
    Object {
      "attrs": Object {
        "key": "t-3",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "You probably build websites and think your ðŸ’©  is special.",
              },
              Object {
                "attrs": Object {
                  "key": "t-0-br-0",
                },
                "children": Array [],
                "tagName": "br",
              },
              Object {
                "key": "t-0-line-0",
                "text": "You think your 13 megabyte parallax-ative home page is going to get you",
              },
            ],
            "key": "t-0",
          },
          Object {
            "attrs": Object {
              "key": "t-1",
            },
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "key": "t-1-line-first",
                    "text": " some fucking Awwward banner y",
                  },
                ],
                "key": "t-1",
              },
            ],
            "tagName": "strong",
          },
          Object {
            "children": Array [
              Object {
                "key": "t-2-line-first",
                "text": "ou can glue to the top corner of your site. You think your 40-pound jQuery file and 83 polyfills give IE7 a boner because it finally has box-shadow. Wrong, motherfucker. Let me describe your perfect-ass website:",
              },
            ],
            "key": "t-2",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-4",
      },
      "children": Array [
        Array [
          Object {
            "attrs": Object {
              "key": "t-0",
            },
            "children": Array [
              Array [
                Object {
                  "attrs": Object {
                    "key": "t-0",
                  },
                  "children": Array [
                    Array [
                      Object {
                        "children": Array [
                          Object {
                            "key": "t-0-line-first",
                            "text": "Shit's lightweight and loads fast",
                          },
                        ],
                        "key": "t-0",
                      },
                    ],
                  ],
                  "tagName": "p",
                },
                Object {
                  "attrs": Object {
                    "key": "t-1",
                  },
                  "children": Array [
                    Array [
                      Object {
                        "attrs": Object {
                          "key": "t-0",
                        },
                        "children": Array [
                          Array [
                            Object {
                              "attrs": Object {
                                "key": "t-0",
                              },
                              "children": Array [
                                Array [
                                  Object {
                                    "children": Array [
                                      Object {
                                        "key": "t-0-line-first",
                                        "text": "Looks the same in all your shitty browsers",
                                      },
                                    ],
                                    "key": "t-0",
                                  },
                                ],
                              ],
                              "tagName": "p",
                            },
                          ],
                        ],
                        "tagName": "li",
                      },
                      Object {
                        "attrs": Object {
                          "key": "t-1",
                        },
                        "children": Array [
                          Array [
                            Object {
                              "attrs": Object {
                                "key": "t-0",
                              },
                              "children": Array [
                                Array [
                                  Object {
                                    "children": Array [
                                      Object {
                                        "key": "t-0-line-first",
                                        "text": "Fits on all your shitty screens",
                                      },
                                    ],
                                    "key": "t-0",
                                  },
                                ],
                              ],
                              "tagName": "p",
                            },
                            Object {
                              "attrs": Object {
                                "key": "t-1",
                              },
                              "children": Array [
                                Array [
                                  Object {
                                    "attrs": Object {
                                      "key": "t-0",
                                    },
                                    "children": Array [
                                      Array [
                                        Object {
                                          "attrs": Object {
                                            "key": "t-0",
                                          },
                                          "children": Array [
                                            Array [
                                              Object {
                                                "children": Array [
                                                  Object {
                                                    "key": "t-0-line-first",
                                                    "text": "The motherfucker's accessible to every asshole that visits your site",
                                                  },
                                                ],
                                                "key": "t-0",
                                              },
                                            ],
                                          ],
                                          "tagName": "p",
                                        },
                                      ],
                                    ],
                                    "tagName": "li",
                                  },
                                ],
                              ],
                              "tagName": "ul",
                            },
                          ],
                        ],
                        "tagName": "li",
                      },
                    ],
                  ],
                  "tagName": "ul",
                },
              ],
            ],
            "tagName": "li",
          },
          Object {
            "attrs": Object {
              "key": "t-1",
            },
            "children": Array [
              Array [
                Object {
                  "attrs": Object {
                    "key": "t-0",
                  },
                  "children": Array [
                    Array [
                      Object {
                        "children": Array [
                          Object {
                            "key": "t-0-line-first",
                            "text": "Shit's legible and gets your fucking point across (if you had one instead of just 5mb pics of hipsters drinking coffee)",
                          },
                        ],
                        "key": "t-0",
                      },
                    ],
                  ],
                  "tagName": "p",
                },
              ],
            ],
            "tagName": "li",
          },
        ],
      ],
      "tagName": "ul",
    },
    Object {
      "attrs": Object {
        "key": "t-5",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Well guess what, motherfucker:",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h3",
    },
    Object {
      "attrs": Object {
        "key": "t-6",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "You. Are. Over-designing. Look at this shit. It's a motherfucking website. Why the fuck do you need to animate a fucking trendy-ass banner flag when I hover over that useless piece of shit? You spent hours on it and added 80 kilobytes to your fucking site, and some motherfucker jabbing at it on their iPad with fat sausage fingers will never see that shit. Not to mention blind people will never see that shit, but they don't see any of your shitty shit.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-7",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "You never knew it, but this is your perfect website. Here's why.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-8",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "It's fucking lightweight",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h2",
    },
    Object {
      "attrs": Object {
        "key": "t-9",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "This entire page weighs less than the gradient-meshed facebook logo on your fucking Wordpress site. Did you seriously load 100kb of jQuery UI just so you could animate the fucking background color of a div? You loaded all 7 fontfaces of a shitty webfont just so you could say \\"Hi.\\" at 100px height at the beginning of your site? ",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-10",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "You piece of shit.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-11",
      },
      "children": Array [
        Array [
          Object {
            "attrs": Object {
              "key": "t-0",
            },
            "children": Array [
              Array [
                Object {
                  "attrs": Object {
                    "key": "t-0",
                  },
                  "children": Array [
                    Array [
                      Object {
                        "attrs": Object {
                          "href": "/blog/11924676",
                          "key": "t-0",
                        },
                        "children": Array [
                          Array [
                            Object {
                              "key": "p-0",
                              "text": "New feature: Media Area granular permissions",
                            },
                          ],
                        ],
                        "tagName": "a",
                      },
                      Object {
                        "attrs": Object {
                          "href": "/blog/11924676",
                          "key": "t-1",
                        },
                        "children": Array [
                          Array [
                            Object {
                              "children": Array [
                                Object {
                                  "key": "t-0-line-first",
                                  "text": "Foobar",
                                },
                              ],
                              "key": "t-0",
                            },
                          ],
                        ],
                        "tagName": "a",
                      },
                    ],
                  ],
                  "tagName": "p",
                },
              ],
            ],
            "tagName": "li",
          },
          Object {
            "attrs": Object {
              "key": "t-1",
            },
            "children": Array [
              Array [
                Object {
                  "attrs": Object {
                    "key": "t-0",
                  },
                  "children": Array [
                    Array [
                      Object {
                        "children": Array [
                          Object {
                            "key": "t-0-line-first",
                            "text": "",
                          },
                        ],
                        "key": "t-0",
                      },
                    ],
                  ],
                  "tagName": "p",
                },
              ],
            ],
            "tagName": "li",
          },
        ],
      ],
      "tagName": "ul",
    },
    Object {
      "attrs": Object {
        "key": "t-12",
        "src": "https://www.datocms-assets.com/205/1606492977-oberlo-cover.jpg",
      },
      "children": Array [],
      "tagName": "img",
    },
    Object {
      "attrs": Object {
        "key": "t-13",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "It's responsive",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h2",
    },
    Object {
      "attrs": Object {
        "key": "t-14",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "You dumbass. You thought you needed media queries to be responsive, but no. Responsive means that it responds to whatever motherfucking screensize it's viewed on. This site doesn't care if you're on an iMac or a motherfucking Tamagotchi.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-15",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "It fucking works",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h2",
    },
    Object {
      "attrs": Object {
        "key": "t-16",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Look at this shit. You can read it ... that is, if you can read, motherfucker. It makes sense. It has motherfucking hierarchy. It's using HTML5 tags so you and your bitch-ass browser know what the fuck's in this fucking site. That's semantics, motherfucker.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-17",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "It has content on the fucking screen. Your site has three bylines and link to your dribbble account, but you spread it over 7 full screens and make me click some bobbing button to show me how cool the jQuery ScrollTo plugin is.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-18",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Cross-browser compatibility? Load this motherfucker in IE6. I fucking dare you.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-19",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "This is a website. ",
              },
            ],
            "key": "t-0",
          },
          Object {
            "attrs": Object {
              "key": "t-1",
            },
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "key": "t-1-line-first",
                    "text": "Look at it.",
                  },
                ],
                "key": "t-1",
              },
            ],
            "tagName": "mark",
          },
          Object {
            "children": Array [
              Object {
                "key": "t-2-line-first",
                "text": " You've never seen one before.",
              },
            ],
            "key": "t-2",
          },
        ],
      ],
      "tagName": "h2",
    },
    Object {
      "attrs": Object {
        "key": "t-20",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Like the man who's never grown out his beard has no idea what his true natural state is, you have no fucking idea what a website is. All you have ever seen are shitty skeuomorphic bastardizations of what should be text communicating a fucking message. This is a real, naked website. Look at it. It's fucking beautiful.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-21",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Yes, this is fucking satire, you fuck",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h3",
    },
    Object {
      "attrs": Object {
        "key": "t-22",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "I'm not actually saying your shitty site should look like this. What I'm saying is that all the problems we have with websites are ones we create ourselves. Websites aren't broken by default, they are functional, high-performing, and accessible. You break them. You son-of-a-bitch.",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "p",
    },
    Object {
      "attrs": Object {
        "key": "t-23",
      },
      "children": Array [
        Array [
          Object {
            "attrs": Object {
              "key": "t-0",
            },
            "children": Array [
              Array [
                Object {
                  "children": Array [
                    Object {
                      "key": "t-0-line-first",
                      "text": "\\"Good design is as little design as possible.\\" ",
                    },
                  ],
                  "key": "t-0",
                },
              ],
            ],
            "tagName": "p",
          },
          Object {
            "attrs": Object {
              "key": "footer",
            },
            "children": Array [
              "some German motherfucker",
            ],
            "tagName": "footer",
          },
        ],
      ],
      "tagName": "blockquote",
    },
    Object {
      "attrs": Object {
        "key": "t-24",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "Epilogue",
              },
            ],
            "key": "t-0",
          },
        ],
      ],
      "tagName": "h2",
    },
    Object {
      "attrs": Object {
        "key": "t-25",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "From the philosophies expressed (poorly) above, ",
              },
            ],
            "key": "t-0",
          },
          Object {
            "attrs": Object {
              "href": "http://txti.es",
              "key": "t-1",
              "rel": "nofollow",
              "target": "_blank",
            },
            "children": Array [
              Array [
                Object {
                  "children": Array [
                    Object {
                      "key": "t-0-line-first",
                      "text": "txti",
                    },
                  ],
                  "key": "t-0",
                },
              ],
            ],
            "tagName": "a",
          },
          Object {
            "children": Array [
              Object {
                "key": "t-2-line-first",
                "text": " was created. You should try it today to make your own motherfucking websites.",
              },
            ],
            "key": "t-2",
          },
        ],
      ],
      "tagName": "p",
    },
  ],
  "key": "t-0",
}
`;

exports[`render unique keys 1`] = `
Object {
  "children": Array [
    Object {
      "attrs": Object {
        "key": "t-0",
      },
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "key": "t-0-line-first",
                "text": "text two\\\\n",
              },
            ],
            "key": "t-0",
          },
          Object {
            "children": Array [
              Object {
                "key": "t-1-line-first",
                "text": "text two\\\\n",
              },
            ],
            "key": "t-1",
          },
        ],
      ],
      "tagName": "p",
    },
  ],
  "key": "t-0",
}
`;
